### PR TITLE
Automatically decide which translations are finished

### PR DESCRIPTION
**Pull request by frosch123**
## Motivation / Problem

Currently OpenTTD distinguishes finished and unfinished translations by putting the translation files into different folders.
* Moving languages from one category to the other is a big hassle. (esp. in coordination with eints)
* Unfinished languages are not compiled, and not part of the nightly or any releases.
* There is no easy way for translators to test unfinished languages.

## Description

The manual sorting into "finished" and "unfinished" is dropped.
* All languages are compiled.
* The language selection dropdown displays a completion percentage for each language (not shown for mostly complete languages, with less than missing_strings_threshold missing strings)
* A language is automatically treated as "unfinished", if the translation is less than 75% complete. (quite a low threshold actually)
    * Unfinished languages are never auto-picked by OpenTTD, even if they match the locale.
    * In release builds, unfinished languages are not offered in the Game Options.
    * In nightly builds, all languages are offered in Game Options.
    * In both release and nightly builds, unfinished translations can be used by editing openttd.cfg

## Limitations

* After this PR, all files from /src/lang/unfinished should be moved to /src/lang. But since this requires coordination with eints (for the last time), it is not part of this PR.
* I considered not building and bundling unfinished translations for release builds, but this seems to lead to build-dependency hell: suddenly the strgen run depends on findversion, so translations are built on every run.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Comment by LordAro** [comment]
Does this change which translations are considered "incomplete" at all?

**Comment by frosch123** [comment]
From the languages, that are currently considered "unfinished":
* Frisian will make it to "finished" with 82%. (after all files from /src/lang/unfinished are moved to /src/lang, see "limitations")
* Persian is very close at 73%.

Of the languages, that are currently considered "finished", all remain as "finished", but some are close to drop out:
* Esperanto: 78%
* Faroese: 81%
* Malay: 84%
* Icelandic: 86%
* Tamil: 87%
* All other languages are at ≥90% completion.

File: src/settings_gui.cpp [hunk 1]
```diff
@@ -229,6 +231,7 @@ struct GameOptionsWindow : Window {
 						 * entries in the dropdown list. */
 						item->SetParamStr(0, _languages[i].name);
 					}
+					item->SetParam(1, (LANGUAGE_TOTAL_STRINGS - _languages[i].missing) * 100 / LANGUAGE_TOTAL_STRINGS);
```
Comment (LordAro) [inline src/settings_gui.cpp:236]: Are there no issues with setting this parameter then not using it (in the case of `STR_JUST_RAW_STRING` ) ?

File: src/settings_gui.cpp [hunk 2]
```diff
@@ -217,7 +218,8 @@ struct GameOptionsWindow : Window {
 
 			case WID_GO_LANG_DROPDOWN: { // Setup interface language dropdown
 				for (uint i = 0; i < _languages.size(); i++) {
-					auto item = new DropDownListParamStringItem(STR_JUST_RAW_STRING, i, false);
+					if (IsReleasedVersion() && !_languages[i].IsReasonablyFinished()) continue;
+					auto item = new DropDownListParamStringItem(IsReleasedVersion() || _languages[i].missing < _settings_client.gui.missing_strings_threshold ? STR_JUST_RAW_STRING : STR_GAME_OPTIONS_LANGUAGE_PERCENTAGE, i, false);
```
Comment (LordAro) [inline src/settings_gui.cpp]: This conditional is complex/long enough that I think it should probably be broken into its own variable

File: src/strings.cpp [hunk 3]
```diff
@@ -1718,6 +1718,15 @@ bool LanguagePackHeader::IsValid() const
 	       StrValid(this->digit_decimal_separator,        lastof(this->digit_decimal_separator));
 }
 
+/**
+ * Check whether a translation is sufficiently finished
+ * to offer it to the public.
```
Comment (LordAro) [inline src/strings.cpp]: Perhaps also say what sufficiently is?

Also, weird linebreak

File: src/settings_gui.cpp (hunk 1 above)
Comment (frosch123) [reply src/settings_gui.cpp:236]: RAW_STRING parameters are only borrowed pointers. The caller has to make sure, they remain valid, as long as anyone processed the string.
In this case the optional parameter is a NUM, not a RAW_STRING.

**Review by rubidium42 — APPROVED** [review]

**Pull request merged**